packages/ab-testing/src/index.ts

```diff
@@ -25,43 +25,48 @@ function getModuloValue(experiment: string, userId: number): number {
     return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
 }
 
-export class Experiments {
-    config: ABTestingConfig;
-    userId: number;
-    userProfile: { [s: string]: string };
-
-    constructor(config: ABTestingConfig, userId: number, userProfile: { [s: string]: string }) {
-        this.config = config;
-        this.userId = userId;
-        this.userProfile = userProfile;
-    }
-
-    getCohort = (experimentName: string): string => {
-        const experimentConfig = this.config.experiments.find(e => e.name === experimentName);
-        if (!experimentConfig) {
-            console.error(`unrecognized ab testing experiment name: ${experimentName}`);
-            return 'control';
-        }
-        const userSegmentNum = getModuloValue(experimentName, this.userId);
-        let allocatedCohort = 'control';
-        for (const cohort of experimentConfig.cohorts) {
-            if (cohort.force_include) {
-                for (const key in cohort.force_include) {
-                    if (cohort.force_include[key].includes(this.userProfile[key])) {
-                        return cohort.name;
-                    }
+function matchUserCohort(
+    experimentConfig: Experiment,
+    userId: number,
+    userProfile: { [s: string]: string }
+): string {
+    const userSegmentNum = getModuloValue(experimentConfig.name, userId);
+    let allocatedCohort = 'control';
+    for (const cohort of experimentConfig.cohorts) {
+        if (cohort.force_include) {
+            for (const key in cohort.force_include) {
+                if (cohort.force_include[key].includes(userProfile[key])) {
+                    return cohort.name;
                 }
             }
-            if (allocatedCohort === 'control') {
-                for (const allocation of cohort.allocation || []) {
-                    if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
-                        allocatedCohort = cohort.name;
-                    }
+        }
+        if (allocatedCohort === 'control') {
+            for (const allocation of cohort.allocation || []) {
+                if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
+                    allocatedCohort = cohort.name;
                 }
             }
         }
-        return allocatedCohort;
-    };
+    }
+    return allocatedCohort;
+}
+
+export class Experiments {
+    matchedCohorts: { [experimentName: string]: string };
+
+    constructor(config: ABTestingConfig, userId: number, userProfile: { [s: string]: string }) {
+        this.matchedCohorts = {};
+        for (const experimentConfig of config.experiments) {

```

Comment:
Sorry I think I've asked for this change; but I'm not sure this is good anymore.

If we end up with a lot of experiments, we'll precalculate them all even though only one or 2 might run into a single page view. Then we'd end up wasting more time on load.

I feel like on-load performance is probably more important than update performance. So if we need to choose between the 2, I think the current implementation will be better.

We could progressively cache each calls to `getCohort()` to improve performance and skip recalculation. Precomputing all doesn't seems like a good solution here - so I think we need to make a call as to whether caching getCohort calls is important enough to add extra complexity.

Reply:
I'm fine with progressive calculation. It's not so complicated, we can add it right now.